syntax = "proto3";

package freqsearch.v1;

option go_package = "github.com/saltfish/freqsearch/go-backend/pkg/pb/freqsearch/v1;freqsearchv1";

import "google/protobuf/timestamp.proto";
import "freqsearch/v1/common.proto";

// Classification tags for a strategy (generated by LLM analysis)
message StrategyTags {
  repeated string strategy_type = 1;   // e.g., ["trend_following", "momentum"]
  string risk_level = 2;               // "low", "medium", "high"
  string trading_style = 3;            // "scalping", "intraday", "swing", "position"
  repeated string indicators = 4;      // Detected indicators from code
  repeated string market_regime = 5;   // "trending", "ranging", "volatile" - added by Analyst
}

// Trading strategy entity
message Strategy {
  string id = 1;
  string name = 2;
  string code = 3;              // Full Python source code
  string code_hash = 4;         // SHA256 hash for deduplication
  optional string parent_id = 5; // Parent strategy ID (for lineage tracking)
  int32 generation = 6;         // Generation number in optimization tree
  string description = 7;
  StrategyMetadata metadata = 8;
  StrategyTags tags = 11;       // Classification tags (LLM-generated)
  google.protobuf.Timestamp created_at = 9;
  google.protobuf.Timestamp updated_at = 10;
}

// Extracted strategy configuration metadata
message StrategyMetadata {
  string timeframe = 1;                  // e.g., "5m", "1h"
  repeated string indicators = 2;        // e.g., ["RSI", "MACD", "EMA"]
  double stoploss = 3;                   // e.g., -0.12 for 12% loss
  bool trailing_stop = 4;
  double trailing_stop_positive = 5;
  double trailing_stop_positive_offset = 6;
  map<string, double> minimal_roi = 7;   // e.g., {"0": 0.1, "60": 0.05}
  int32 startup_candle_count = 8;
}

// Strategy with its best performance metrics
message StrategyWithMetrics {
  Strategy strategy = 1;
  StrategyPerformanceMetrics best_result = 2;
  int32 backtest_count = 3;
}

// Aggregated performance metrics for a strategy
message StrategyPerformanceMetrics {
  double sharpe_ratio = 1;
  double sortino_ratio = 2;
  double profit_pct = 3;
  double max_drawdown_pct = 4;
  int32 total_trades = 5;
  double win_rate = 6;
  double profit_factor = 7;
}

// ----- Strategy Service RPCs -----

message CreateStrategyRequest {
  string name = 1;
  string code = 2;
  optional string parent_id = 3;
  string description = 4;
  StrategyTags tags = 5;        // Classification tags
}

message CreateStrategyResponse {
  Strategy strategy = 1;
}

message GetStrategyRequest {
  string id = 1;
}

message GetStrategyResponse {
  Strategy strategy = 1;
}

message SearchStrategiesRequest {
  optional string name_pattern = 1;   // SQL LIKE pattern
  optional double min_sharpe = 2;
  optional double min_profit_pct = 3;
  optional int32 min_trades = 4;
  optional double max_drawdown_pct = 5;
  PaginationRequest pagination = 6;
  string order_by = 7;                // "sharpe", "profit", "created_at"
  bool ascending = 8;
}

message SearchStrategiesResponse {
  repeated StrategyWithMetrics strategies = 1;
  PaginationResponse pagination = 2;
}

message GetStrategyLineageRequest {
  string strategy_id = 1;
  int32 depth = 2;  // How many generations to traverse
}

message GetStrategyLineageResponse {
  repeated StrategyLineageNode lineage = 1;
}

message StrategyLineageNode {
  Strategy strategy = 1;
  optional StrategyPerformanceMetrics metrics = 2;
  repeated StrategyLineageNode children = 3;
}

message DeleteStrategyRequest {
  string id = 1;
}

message DeleteStrategyResponse {
  bool success = 1;
}
